### PR TITLE
stdlib: prevent static builds on Darwin

### PR DESCRIPTION
**Pull request by compnerd**
Static builds of the standard library and the SDK overlay are
unsupported post ABI stability.  Disallow building of the standard
library and SDK overlay statically for Darwin platforms.

Replace this paragraph with a description of your changes and rationale. Provide links to external references/discussions if appropriate.

Resolves [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN).

### CODE REVIEW

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
CC: @shahmishal @brentdax @Rostepher @gottesmm @jrose-apple

File: stdlib/CMakeLists.txt [hunk 1]
```diff
@@ -77,7 +77,12 @@ if(SWIFT_BUILD_DYNAMIC_STDLIB)
   list(APPEND SWIFT_STDLIB_LIBRARY_BUILD_TYPES SHARED)
 endif()
 if(SWIFT_BUILD_STATIC_STDLIB)
-  list(APPEND SWIFT_STDLIB_LIBRARY_BUILD_TYPES STATIC)
+  list_interect("${SWIFT_APPLE_PLATFORMS}" "${SWIFT_SDKS}" building_darwin_sdks)
+  if(building_dawin_sdks)
```
Comment (DougGregor) [inline stdlib/CMakeLists.txt]: Typo here "building_dawin_sdks" vs. "building_darwin_sdks".

File: stdlib/public/Darwin/CMakeLists.txt [hunk 2]
```diff
@@ -5,7 +5,12 @@ if(SWIFT_BUILD_DYNAMIC_SDK_OVERLAY)
   list(APPEND SWIFT_SDK_OVERLAY_LIBRARY_BUILD_TYPES SHARED)
 endif()
 if(SWIFT_BUILD_STATIC_SDK_OVERLAY)
-  list(APPEND SWIFT_SDK_OVERLAY_LIBRARY_BUILD_TYPES STATIC)
+  list_interect("${SWIFT_APPLE_PLATFORMS}" "${SWIFT_SDKS}" building_darwin_sdks)
+  if(building_dawin_sdks)
```
Comment (DougGregor) [inline stdlib/public/Darwin/CMakeLists.txt]: Same typo.

**Review by DougGregor — CHANGES_REQUESTED** [review]

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
Funny, I just filed a bug for this. https://bugs.swift.org/browse/SR-11157

**Comment by compnerd** [comment]
Yes, it does make that impossible, but, really, that is already pretty badly broken.  The current model doesn't really work (CMake just is not designed that way) - e.g. `-install_name` being used on ELF and `-soname` being used for MachO?!  We have a ton of hacks in place and we keep piling more on.  I think that we need to just change to proper cross-compilation (nearly all the pieces are now in place).  At that point, that will have different CMake sub-builds for each one.  This is better long term.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by brentdax** [comment]
I think @shahmishal needs to be part of this conversation, because if we can't use the same presets on Darwin and other platforms, that's going to impact the way CI is configured.

**Comment by shahmishal** [comment]
@brentdax @compnerd Thanks for bring this to my attention, if we see build failures on CI because we are still using the flag on macOS please let me know and I will update the preset.

**Comment by brentdax** [comment]
Looking at build-presets.ini, the following apparently-Darwin presets still use this flag:

* mixin_buildbot_trunk_base
* buildbot,tools=RA-flto,stdlib=RDA
* LLDB_Swift_DebugAssert_with_devices
* LLDB_Swift_ReleaseAssert_with_devices

We should probably remove the flag from these.

**Comment by compnerd** [comment]
Beyond the update of the presets, the question is, should we even permit building the standard library/SDK overlay for Darwin in static mode?

**Comment by brentdax** [comment]
We think not permitting it is fine. Mishal says that we (should) always use separate presets for Linux, so preset sharing isn't a concern.

**Comment by compnerd** [comment]
@swift-ci please test macOS platform

**Comment by compnerd** [comment]
@swift-ci please test macOS platform

**Comment by compnerd** [comment]
Sounds like all the concerns have been addressed, I'm going to go ahead and merge this.  As @shahmishal said, if we run into cases where the preset is using static builds, we will update the preset and sharing is not a concern.

**Comment by shahmishal** [comment]
https://github.com/apple/swift/pull/26343

**Pull request merged**